core/win32/module_shared.c

```diff
@@ -1326,6 +1326,41 @@ read_remote_maybe64(HANDLE process, uint64 addr, size_t bufsz, void *buf)
         num_read == bufsz;
 }
 
+uint64
+find_remote_dll_base(HANDLE phandle, bool find64bit, char *dll_name)
+{
+    MEMORY_BASIC_INFORMATION64 mbi;
+    uint64 got;
+    NTSTATUS res;
+    uint64 addr = 0;
+    char name[MAXIMUM_PATH];
+    do {
+        res = remote_query_virtual_memory_maybe64(phandle, addr, &mbi, sizeof(mbi), &got);
+        if (got != sizeof(mbi) || !NT_SUCCESS(res))
+            break;
+#    if VERBOSE
+        print_file(STDERR, "0x%I64x-0x%I64x type=0x%x state=0x%x\n", mbi.BaseAddress,
+                   mbi.BaseAddress + mbi.RegionSize, mbi.Type, mbi.State);
+#    endif
+        if (mbi.Type == MEM_IMAGE && mbi.BaseAddress == mbi.AllocationBase) {
+            bool is_64;
+            if (get_remote_dll_short_name(phandle, mbi.BaseAddress, name,
+                                          BUFFER_SIZE_ELEMENTS(name), &is_64)) {
+#    if VERBOSE
+                print_file(STDERR, "found |%s| @ 0x%I64x 64=%d\n", name, mbi.BaseAddress,
+                           is_64);
+#    endif
+                if (_strcmpi(name, dll_name) == 0 && BOOLS_MATCH(find64bit, is_64))
```

Comment:
> It also compares with _strcmpi because kernel32 was loaded upper case.
>  When I used strcmp and did not find anything, find_remote_dll_base would go into an infinite loop - I did not check the root cause of it.

You mean the outer loop never terminated?  I'm not seeing how that could happen -- mbi.RegionSize being incorrect??  But we have many other loops like this.

> It does not cause any issues now, because it is only used with ntdll and kernel32 which are always loaded. But now that it accepts general DLLs, I think it should be fixed.

No kernel32 is not always loaded: early on in the life of a process there's only ntdll, and DR's earliest injection takes over well before kernel32 is loaded, so you could imagine this function being called then by DR code.  (And I think it's possible to create a process that never uses kernel32 but a pathological case for sure.)

But maybe not worth effort at this point

Reply:
>You mean the outer loop never terminated? I'm not seeing how that could happen -- mbi.RegionSize being incorrect?? But we have many other loops like this.

Output was something like this:

```
...
found |wow64.dll| @ 0x7ffca0d40000 64=1
0x7ffca0d41000-0x7ffca0d7a000 type=0x1000000 state=0x1000
0x7ffca0d7a000-0x7ffca0d8f000 type=0x1000000 state=0x1000
0x7ffca0d8f000-0x7ffca0d91000 type=0x1000000 state=0x1000
0x7ffca0d91000-0x7ffca0d99000 type=0x1000000 state=0x1000
0x7ffca0d99000-0x7ffca1ad0000 type=0x0 state=0x10000
0x7ffca1ad0000-0x7ffca1ad1000 type=0x1000000 state=0x1000
found |ntdll.dll| @ 0x7ffca1ad0000 64=1
0x7ffca1ad1000-0x7ffca1bec000 type=0x1000000 state=0x1000
0x7ffca1bec000-0x7ffca1c34000 type=0x1000000 state=0x1000
0x7ffca1c34000-0x7ffca1c35000 type=0x1000000 state=0x1000
0x7ffca1c35000-0x7ffca1c37000 type=0x1000000 state=0x1000
0x7ffca1c37000-0x7ffca1c40000 type=0x1000000 state=0x1000
0x7ffca1c40000-0x7ffca1cc5000 type=0x1000000 state=0x1000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
...
...
...
```

>No kernel32 is not always loaded: early on in the life of a process there's only ntdll, and DR's earliest injection takes over well before kernel32 is loaded, so you could imagine this function being called then by DR code. (And I think it's possible to create a process that never uses kernel32 but a pathological case for sure.)
>
>But maybe not worth effort at this point

I see. But for now we only use kernel32 for attach, so maybe it is ok to assume it is loaded? (though clearly not the best)

Reply:
I would add a XXX or TODO or even FIXME comment to that loop explaining the infinite loop you saw.

Reply:
I'll do that.

In the meanwhile, I added a print of the address:
```
...
found |ntdll.dll| @ 0x7ffca1ad0000 64=1
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1ad1000
0x7ffca1ad1000-0x7ffca1bec000 size=0x11b000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1bec000
0x7ffca1bec000-0x7ffca1c34000 size=0x48000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c34000
0x7ffca1c34000-0x7ffca1c35000 size=0x1000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c35000
0x7ffca1c35000-0x7ffca1c37000 size=0x2000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c37000
0x7ffca1c37000-0x7ffca1c40000 size=0x9000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c40000
0x7ffca1c40000-0x7ffca1cc5000 size=0x85000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1cc5000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x7fffffff0000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x80035e31b000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x8006bc646000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x800a1a971000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
...
```

So something weird happens when the address passes 0x800000000000. 
Maybe just add
```
if (mbi.BaseAddress + mbi.RegionSize < addr)
    break;
```
WDYT?

Reply:
Is this infinite loop only happening for 32-bit (WOW64)?  The drrun and the target are the same bitwidth?

Reply:
Yes, I only tested on a 64 bit machine: it only happens for 32-bit (both the drrun and infloop).

Reply:
Done

Reply:
I do remember some weird Windows kernel behavior with memory query walks but not quite this one. 
 The one I'm thinking of is the AllocationBase: https://github.com/DynamoRIO/drmemory/issues/2328#issuecomment-739454412